streams/src/test/java/org/apache/kafka/streams/integration/WindowedChangelogRetentionIntegrationTest.java

```diff
@@ -0,0 +1,326 @@
+        final Duration windowSize = DEFAULT_RETENTION.minus(grace).minus(Duration.ofMillis(500));
+        final Duration expectedRetention = windowSize.plus(grace);
+        runAndVerifyTimeWindows(TimeWindows.of(windowSize).grace(grace), null, expectedRetention);
+    }
+
+    @Test
+    public void timeWindowedChangelogShouldHaveUserSpecifiedRetentionIfUserSpecifiedRetentionEvenIfLessThanDefaultRetention() throws Exception {
+        final Duration grace = Duration.ofHours(6);
+        final Duration windowSize = DEFAULT_RETENTION.minus(grace).minus(Duration.ofHours(1));
+        final Duration userSpecifiedRetention = windowSize.plus(grace).plus(Duration.ofHours(1));
+        final Duration expectedRetention = userSpecifiedRetention;
+        runAndVerifyTimeWindows(TimeWindows.of(windowSize).grace(grace), userSpecifiedRetention, expectedRetention);
+    }
+
+    private void runAndVerifyTimeWindows(final Windows<TimeWindow> window,
+                                         final Duration userSpecifiedRetention,
+                                         final Duration expectedRetention) throws Exception {
+        final String storeName = "windowed-store";
+        final Serde<Windowed<String>> windowedSerde = WindowedSerdes.timeWindowedSerdeFrom(String.class, window.size());
+        groupedStream.windowedBy(window)
+            .count(userSpecifiedRetention != null
+                ? Materialized.<String, Long, WindowStore<Bytes, byte[]>>as(storeName).withRetention(userSpecifiedRetention)
+                : Materialized.as(storeName))
+            .toStream()
+            .to(OUTPUT_TOPIC, Produced.with(windowedSerde, Serdes.Long()));
+
+        startStreams();
+
+        verifyChangelogRetentionOfWindowedStore(storeName, expectedRetention);
+    }
+
+    @Test
+    public void sessionWindowedChangelogShouldHaveRetentionOfGapIfGapLargerThanDefaultRetention() throws Exception {
+        final Duration gap = DEFAULT_RETENTION.plus(Duration.ofHours(1));
+        final Duration expectedRetention = gap;
+        runAndVerifySessionWindows(SessionWindows.with(gap), null, expectedRetention);
+    }
+
+    @Test
+    public void sessionWindowedChangelogShouldHaveDefaultRetentionIfGapLessThanDefaultRetention() throws Exception {
+        final Duration gap = DEFAULT_RETENTION.minus(Duration.ofHours(1));
+        final Duration expectedRetention = DEFAULT_RETENTION;
+        runAndVerifySessionWindows(SessionWindows.with(gap), null, expectedRetention);
+    }
+
+    @Test
+    public void sessionWindowedChangelogShouldHaveRetentionOfGapPlusGraceIfGapPlusGraceLessThanDefaultRetention() throws Exception {
+        final Duration grace = Duration.ofHours(1);
+        final Duration gap = DEFAULT_RETENTION.minus(grace).minus(Duration.ofHours(1));
+        final Duration expectedRetention = gap.plus(grace);
+        runAndVerifySessionWindows(SessionWindows.with(gap).grace(grace), null, expectedRetention);
+    }
```

Comment:
With 3.0 we do not use the default retention if gap + grace is less than the default retention. I think that is fine because a changelog topic that is created with 2.8 has a larger retention time than a changelog topic created with 3.0. \cc @mjsax